### PR TITLE
Feature/modal routing

### PR DESCRIPTION
**Pull request by nielslyngsoe**
Modal routing feature.

Enables registering a modal, so it can be opened via URL.

### Manifest

```json
{
	"type": "modal",
	"alias": "Our.Modal.SomethingPicker",
	"name": "My Something Picker Modal",
	"loader": "./my-something-picker-modal-element.js",
},
```

### Modal Token

For type safety we recommend that you make a Modal Token, Its posible to go without.
The Modal Token binds the Modal Type to the Modal Data Type and Modal Result Type.

``

```ts
import { ModalToken } from '@umbraco-cms/element';

export type OurSomethingPickerModalData = {
	key: string | null;
};

export type OurSomethingPickerModalResult = {
	key: string;
};

export const MY_SOMETHING_PICKER_MODAL = new UmbModalToken<UmbLinkPickerModalData, UmbLinkPickerModalResult>(
	'Our.Modal.SomethingPicker',
	{
		type: 'sidebar',
		size: 'small',
	}
);
```

### Make a modal registration

When registering you need to be aware of the scope of the modal.
A modal registered in a dashboard can be relatively simple.
A modal registered in a property editor needs to become specific for the property and the variant of that property.

Notice we are using a Controller here, this means your element has to be a Controller Host (TODO: Insert link to story about Controller Host also available through the UmbElementMixin)

#### Simple Modal Route Registration

A modal can be registred via the UmbModalRouteRegistrationController. The registration will accept a modal token (or extension alias).

Notice we are using a Controller here, this means your element has to be a Controller Host (TODO: Insert link to story about Controller Host also available through the UmbElementMixin)

```ts
this.myModalRegistration = new UmbModalRouteRegistrationController(this, UMB_LINK_PICKER_MODAL_TOKEN)
	.onSubmit((submitData) => {
		console.log('Modal submitted with data'.submitData);
	})
	.observeRouteBuilder((routeBuilder) => {
		this._modalRouteBuilder = routeBuilder;
	});
```

#### Modal registration for UI as part of a Property Editor

```ts


	@property()
	public set alias(value: string | undefined) {
		this.myModalRegistration.setUniqueIdentifier('propertyAlias', value);
	}

	@property()
	public set variantId(value: string | UmbVariantId | undefined) {
		this.myModalRegistration.setUniqueIdentifier('variantId', value?.toString());
	}

	private _items = [
		{ name: 'Item 1' },
		{ name: 'Item 2' },
		{ name: 'Item 3' },
	]


	constructor() {
		super();

		this.myModalRegistration = new UmbModalRouteRegistrationController(
			this,
			MY_MODAL_TOKEN,
			`:index`,
			new Map([
				['propertyAlias', undefined],
				['variantId', undefined],
			])
		)
		.onSetup((params) => {
			// Get item index:
			const indexParam = params.index;
			if (!indexParam) return false;
			let index: number | null = parseInt(params.index);
			if (Number.isNaN(index)) return false;

			// Use the index to find data:
			let data = null;
			if (index >= 0 && index < this._items.length) {
				data = this._items[index];
			} else {
				// If not then make a new pick:
				index = null;
			}

			return {
				index: index,
				itemData: {
					name: data?.name
				},
			};
		})
		.onSubmit((submitData) => {
			if (!submitData) return;
			this._items[submitData.index] = submitData.itemData;
		})
		.observeRouteBuilder((routeBuilder) => {
			this._modalRouteBuilder = routeBuilder;
		});
	}

	render() {
		return html`
			${this._items?.map((item, index) =>
				html`<uui-button look="placeholder" label="Edit item ${index}" .href=${this._modalRouteBuilder?.({ index })}>Add</uui-button>`
			)}
		`;
	}
```

## Generate the URL to a Modal

The Modal registration has an option to retrive a URL Builder, this is a function that can be used to generate a URL to a modal.

```ts
const modalLink = _modalRouteBuilder?.({ alias: 'my-input-alias' });
```

The modalLink from above could look like: /umbraco/backoffice/my/location/modal/Our.Modal.SomethingPicker/my-input-alias

Notice the Property Editor Modal which has unique parts in its registration will add the property alias and variant id to the URL, so it becomes:
/umbraco/backoffice/my/location/modal/Our.Modal.SomethingPicker/my-property-alias/en-us/my-key-1234

### CODE REVIEW

**Review by iOvergaard — CHANGES_REQUESTED** [review]
Code looks good, but unfortunately I'm getting a "router-slot has already been defined" sort of error when testing on the v13/dev branch. It looks like router-slot gets picked up with all its side-effects in the new libs/router module as well as through the backoffice, thus imported twice and they do not check if a component named "router-slot" already exists.

image

File: libs/router/route.interface.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import type { IRoute } from 'router-slot/model';
+/*
+type Optional<T> = { [P in keyof T]?: T[P] };
+
+type IRouteWithComponentRoute = Exclude<IRoute, IComponentRoute>;
+type IComponentRouteWithOptionalComponent = Omit<IComponentRoute, 'component'> &
+	Optional<Pick<IComponentRoute, 'component'>>;
+
+export type UmbRoute = IRouteWithComponentRoute | IComponentRouteWithOptionalComponent;
+*/
```
[the remark this answers is not in the document]
Comment (nielslyngsoe) [reply libs/router/route.interface.ts]: cool, I have removed these comments. they are not important for now. :-)

File: eslint-local-rules.cjs
[the remark this answers is not in the document]
Comment (nielslyngsoe) [reply eslint-local-rules.cjs:1]: Thanks. Though its not perfect in correction those imports. But it highlights them at least.

**Review by iOvergaard — APPROVED** [review]
It works on the v13/dev branch now and looks good to me.

**Pull request merged**